src/main/java/org/wildfly/security/keystore/KeyStoreWatcher.java

```diff
@@ -0,0 +1,227 @@
+        newWatchedPaths.put(dirPath, pathRegistration);
+        watchedPaths = newWatchedPaths;
+
+        if (watchRequired) {
+            if (watchService == null) {
+                watchService = fileSystem.newWatchService();
+                executor.execute(new EventTaker());
+            }
+            // We use 'create' in addition to 'modify' as updates could be in the form of replacing a file.
+            WatchKey key = dirPath.register(watchService, StandardWatchEventKinds.ENTRY_CREATE, StandardWatchEventKinds.ENTRY_MODIFY);
+            registrations.put(dirPath, key);
+        }
+    }
+
+    synchronized void deRegister(File watchFile, Store keyStore) throws IOException {
+        File canonical = watchFile.getCanonicalFile();
+
+        File parentDir = canonical.getParentFile();
+        String fileName = canonical.getName();
+
+        Path dirPath = parentDir.toPath();
+
+        boolean watchRequired = false;
+        Map<String, List<Store>> pathRegistration = watchedPaths.get(dirPath);
+        List<Store> pathStores = null;
+        if (pathRegistration != null) {
+            pathRegistration = new HashMap<String, List<Store>>(pathRegistration);
+            List<Store> tmpStores = pathRegistration.get(fileName);
+            if (tmpStores != null) {
+                pathStores = new LinkedList<Store>(tmpStores);
+                Iterator<Store> storeIterator = pathStores.iterator();
+                while (storeIterator.hasNext()) {
+                    Store current = storeIterator.next();
+                    if (keyStore == current) {
+                        storeIterator.remove();
+                    }
+                }
+                if (pathStores.isEmpty()) {
+                    pathRegistration.remove(fileName);
+                } else {
+                    pathRegistration.put(fileName, pathStores);
+                }
+            }
+
+            Map<Path, Map<String, List<Store>>> newWatchedPaths = new HashMap<Path, Map<String, List<Store>>>(watchedPaths);
+            if (pathRegistration.isEmpty()) {
+                newWatchedPaths.remove(dirPath);
+                WatchKey key = registrations.remove(dirPath);
+                if (key != null) {
+                    key.cancel();
+                }
+                if (newWatchedPaths.isEmpty()) {
+                    watchService.close();
+                    watchService = null;
+                }
+            } else {
+                newWatchedPaths.put(dirPath, pathRegistration);
+            }
+            watchedPaths = newWatchedPaths;
+        }
+
+    }
+
+    interface Store {
+
+        void modified();
+
+    }
+
+    private class EventTaker implements Runnable {
+
+        @Override
+        public void run() {
+            try {
+                while (watchService != null) {
+                    WatchKey key = watchService.take();
```

Comment:
NPE risk here.  The watchService field should be copied to a local variable before the test in case it is changed to null in another thread between the "while" and the call to .take().